### DISCUSSION TITLE
Don't map values when source value is null (again)

### DISCUSSION THREAD

**Discussion by PaulVrugt** [LuckyPennySoftware/AutoMapper#4333] [Ideas]
So I've been breaking my head over this. If you search for it, the internet is full of people with the same question, and all kinds of clunky solutions have been created.

A common scenario when using automapper combined with EF core, is that we want to update a bunch of properties on an entity, but only update properties that have a value. 

So we have a source type that looks like this:

```c#
public class MutationCar
{
  public int? NumberOfWheels {get;set;}
  public string? Brand {get;set;}
}
```

and a EF core Poco that looks like:
```c#
public class CarPoco
{
  public int NumberOfWheels {get;set;}
  public string Brand {get;set;}
}
```

notice that in the `MutationCar` all properties are nullable, and in the `CarPoco` they are not (because in the database these properties are required.

Now we want to map a `MutationCar` on a `CarPoco` to update the car in the database:

```c#
var existingCar = dbContext.Cars.Find(x => x.cardId = 1);
_mapper.Map(mutationCar, existingCar);
```

To make this work, we have to avoid automapper trying to set destination members when the source member is null. However, we don't want to manually map each property individually and set null mapping properties, so we want to use `ForAllMembers` and we have 2 options to use: `PreCondition` and `Condition`. The problem with `Precondition` is that don't get any reference to the source member or value, so we cannot use this. The `Condition` option seems to work, but as stated in all kinds of earlier issues, it has the limitation when mapping from a nullable to a non-nullable value type, it converts null values of the source member to its default non-nullable variant (for for bool: `null` becomes `false`, for int, `null` becomes `0` etc.).

Seeing the amount of issues for automapper and questions on stackoverflow, there seems to be a demand for a proper solution without having to use custom `ValueMemberResolver` classes. Ideally, any of the below:

- A method on `IMapperExpression` called `SkipNullValues` or something, that accomplishes this behavior
- Some way of accessing the actual source value in the `Condition` call

Now I'm probably willing to contribute to a proper solution for this, but only if it has a realistic change of getting merged. Therefore I would first like to discuss this, or see if there are other better alternatives for a generic solution before starting to create code for this

**Comment by lbargaoanu** [comment]
There are a few generic ways to do this and you're going to have to choose whatever works best for you. If you need help with that, you should ask on SO.

**Reply by PaulVrugt** [reply]
Yeah this is exactly what I'm talking about. All these topics about this issue result in people saying there are ways to do this, that all involve a custom valuememberresolver or mapping properties manually one by one. The current solutions are not very trivial, hence the large amount of questions/discussions about this topic on SO and this repo.

What I'm talking about is to see if we can create a better single generic solution to be able to do this in an easy way.

**Reply by lbargaoanu** [reply]
No, it's easy enough to that in a generic way in your own code.

**Comment by lbargaoanu** [comment]
I think this is better suited for StackOverflow.

**Comment by jbogard** [comment]
I'm a bit hesitant to add any more features to reverse mapping as well. It's a much harder problem to solve so I don't really want to add too many features to it. I'd like to promote just regular code.
